### PR TITLE
Handle dependent context functions

### PR DESCRIPTION
**Pull request by nicolasstucki**
Add `FunctionTypeOfMethod` extractor that matches any kind of function and return its method type.

We use this extractor instead of `ContextFunctionType` to all of
 * `ContextFunctionN[...]`
 * `ContextFunctionN[...] { def apply(using ...): R }` where `R` might be dependent on the parameters.
 * `PolyFunction { def apply(using ...): R }` where `R` might be dependent on the parameters. Currently this one would have at least one erased parameter.

The naming of the extractor follows the idea in https://github.com/lampepfl/dotty/pull/18305#discussion_r1294652027.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 1]
```diff
@@ -1108,6 +1108,22 @@ class Definitions {
     //  - .linkedClass: the ClassSymbol of the enumeration (class E)
     sym.owner.linkedClass.typeRef
 
+  object FunctionTypeOfMethod {
+    /** Matches a `FunctionN[...]`/`ContextFunctionN[...]` or refined `PolyFunction`/`FunctionN[...]`/`ContextFunctionN[...]`.
+     *  Extracts the method type type and apply info.
+     */
+    def unapply(ft: Type)(using Context): Option[MethodOrPoly] = {
+      ft match
+        case RefinedType(parent, nme.apply, mt: MethodOrPoly)
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala:1117]: I'd try to optimize it further by inlining the FunctionOf extractor and specializing it to this use case.
Also, maybe split the `RefinedType` match into two, one for `MethodType` refinements and the other for `PolyType` refinements?

**Review by odersky — CHANGES_REQUESTED** [review]
Otherwise LGTM

**Comment by odersky** [comment]
@nicolasstucki Can you have a look at the test failure? happy to review again when it is green.

**Comment by nicolasstucki** [comment]
@odersky now the CI passed.

**Review by odersky — APPROVED** [review]
This will be very handy to have. Sorry for dropping the ball on the reviews.

**Pull request merged**